include/nuttx/efuse/efuse.h

```diff
@@ -0,0 +1,155 @@
+ * Arguments:   A structure containing the field[], the src memory and the
+ *              amount of bits to write.
+ * Return:      Zero (OK) on success.  Minus one will be returned on failure
+ *              with the errno value set appropriately.
+ */
+
+#define EFUSEIOC_WRITE_FIELD     _EFUSEIOC(0x0002)
+
+/****************************************************************************
+ * Public Types
+ ****************************************************************************/
+
+/* Structure eFuse field */
+
+struct efuse_desc_s
+{
+  uint16_t   bit_offset; /* Bit offset related to beginning of efuse */
+  uint16_t   bit_count;  /* Length of bit field */
+};
+
+/* Type definition for an eFuse field */
+
+typedef struct efuse_desc_s efuse_desc_t;
+
+/* Structs with the parameters passed to the IOCTLs */
+
+struct efuse_param
+{
+  FAR const efuse_desc_t **field;
+  size_t  size;
+  FAR uint8_t *data;
+};
+
+/* This structure provides the "lower-half" driver operations available to
+ * the "upper-half" driver.
+ */
+
+struct efuse_lowerhalf_s;
+struct efuse_ops_s
+{
+  /* Required methods *******************************************************/
+
+  /* Read an EFUSE bit */
+
+  CODE int (*read_field)(FAR struct efuse_lowerhalf_s *lower,
+                         FAR const efuse_desc_t *field[],
+                         FAR uint8_t *data, size_t bit_size);
+
+  /* Write an EFUSE bit */
+
+  CODE int (*write_field)(FAR struct efuse_lowerhalf_s *lower,
+                          FAR const efuse_desc_t *field[],
+                          FAR const uint8_t *data, size_t bit_size);
+
+  /* Any ioctl commands that are not recognized by the "upper-half" driver
+   * are forwarded to the lower half driver through this method.
+   */
+
+  CODE int (*ioctl)(FAR struct efuse_lowerhalf_s *lower, int cmd,
+                    unsigned long arg);
+};
+
+/* This structure provides the publicly visible representation of the
+ * "lower-half" driver state structure.  "lower half" drivers will have an
+ * internal structure definition that will be cast-compatible with this
+ * structure definitions.
+ */
+
+struct efuse_lowerhalf_s
+{
+  /* Publicly visible portion of the "lower-half" driver state structure. */
+
+  FAR const struct efuse_ops_s  *ops;  /* Lower half operations */
+
+  /* The remainder of the structure is used by the "lower-half" driver
+   * for whatever state storage that it may need.
+   */
+
+  void   *upper;                       /* Pointer to watchdog_upperhalf_s */
+};
+
+/****************************************************************************
+ * Public Data
+ ****************************************************************************/
+
+#ifdef __cplusplus
+#define EXTERN extern "C"
+extern "C"
+{
+#else
+#define EXTERN extern
+#endif
+
+/****************************************************************************
+ * Public Function Prototypes
+ ****************************************************************************/
+
```

Comment:
add efuse_register/efuse_unregister prototype here

Reply:
I don't unterstand why lowerhalf require upper pointer?

Reply:
Hi @xiaoxiang781216, saving the upper pointer is important when we have to exchange information inside the ISR handle, in the ESP32 case we don't use interrupts, but the driver should be prepared for other hardware that eventually will use it. Please see the arch/xtensa/src/esp32/esp32_wdt_lowerhalf.c to understand how it is used.

Reply:
Yes, I thought this case before, but if the lower half need callback into upper half in ISR process, the normal approach is pass the upper pointer in set_xxx_callback, e.g.:
```
  CODE int (*start)(FAR struct oneshot_lowerhalf_s *lower,
                    oneshot_callback_t callback, FAR void *arg,
                                                 ^^^^^^^^^^^^^
                    FAR const struct timespec *ts);
```
Or like esp32_wdt_lowerhalf.c case, save it in the private esp32_wdt_lowerhalf_s not public watchdog_lowerhalf_s.

Reply:
Right, I will change it!